File: src/main/java/es/danisales/datune/degrees/scale/DiatonicDegree.java
package es.danisales.datune.degrees.scale;

import es.danisales.datune.degrees.octave.Diatonic;
import es.danisales.datune.interval.IntervalDiatonic;
import es.danisales.utils.MathUtils;
import es.danisales.utils.Utils;
import org.checkerframework.checker.nullness.qual.NonNull;

public enum DiatonicDegree implements ScaleDegree {
	I, II, III, IV, V, VI, VII;

	public static DiatonicDegree from(IntervalDiatonic intervalDiatonic) {
		return values()[intervalDiatonic.ordinal() % DiatonicDegree.values().length];
	}

	/**
	 * Returns the relativedegree with the added interval function
	 *
	 * @param id the id
	 * @return the relativedegree
	 */
	public static DiatonicDegree add(DiatonicDegree diatonicDegree, IntervalDiatonic id) {
		return fromIndex( diatonicDegree.ordinal() + id.ordinal() );
	}

	public static DiatonicDegree sub(DiatonicDegree diatonicDegree, IntervalDiatonic id) {
		return fromIndex( diatonicDegree.ordinal() - id.ordinal() );
	}

	@Override
	public DiatonicDegree getShifted(int i) {
		i = MathUtils.rotativeTrim(i, values().length);
		return values()[i];
	}

	/**
	 * Gets the relativedegree fromIndex an int
	 *
	 * @param n the n
	 * @return the relativedegree
	 */
	private static DiatonicDegree fromIndex(int n) {
		n = limitToOneOctave(n);

		return values()[n];
	}

	private static int limitToOneOctave(int n) {
		n = n % Diatonic.NUMBER;
		if ( n < 0 )
			n += Diatonic.NUMBER;
		return n;
	}

	@NonNull
	@Override
	public DiatonicDegree getNext() {
		int index = (ordinal() + 1) % DiatonicDegree.values().length;
		return values()[index];
	}

	@NonNull
	@Override
	public DiatonicDegree getPrevious() {
		int index = Utils.rotativeTrimLowerOnce(ordinal() - 1, values().length);

		return values()[index];
	}

	@Override
	public String toString() {
		return "Diatonic " + super.toString();
	}
}

File: src/degrees/scale/DiatonicDegree.ts
import { Diatonic } from '../../degrees/Diatonic';
import { Immutables } from '../../common/Immutables';
import { MathUtils } from '../../common/MathUtils';

export class DiatonicDegree {
    public static I;
    public static II;
    public static III;
    public static IV;
    public static V;
    public static VI;
    public static VII;

    private constructor(private _intValue: number) {
    }

    public static fromInt(n: number): DiatonicDegree {
        n = MathUtils.rotativeTrim(n, Diatonic.NUMBER);

        switch (n) {
            case 0: return DiatonicDegree.I;
            case 1: return DiatonicDegree.II;
            case 2: return DiatonicDegree.III;
            case 3: return DiatonicDegree.IV;
            case 4: return DiatonicDegree.V;
            case 5: return DiatonicDegree.VI;
            case 6: return DiatonicDegree.VII;
        }

        throw new Error("Cannot get DiatonicDegree from int: " + n);
    }

    get intValue(): number {
        return this._intValue;
    }

    public hashCode(): string {
        return "dd:" + this._intValue;
    }

    public toString(): string {
        switch (this) {
            case DiatonicDegree.I: return "I";
            case DiatonicDegree.II: return "II";
            case DiatonicDegree.III: return "III";
            case DiatonicDegree.IV: return "IV";
            case DiatonicDegree.V: return "V";
            case DiatonicDegree.VI: return "VI";
            case DiatonicDegree.VII: return "VII";
        }

        throw new Error();
    }

    private static initialize() {
        DiatonicDegree.I = new DiatonicDegree(0);
        DiatonicDegree.II = new DiatonicDegree(1);
        DiatonicDegree.III = new DiatonicDegree(2);
        DiatonicDegree.IV = new DiatonicDegree(3);
        DiatonicDegree.V = new DiatonicDegree(4);
        DiatonicDegree.VI = new DiatonicDegree(5);
        DiatonicDegree.VII = new DiatonicDegree(6);

        Immutables.lockr(DiatonicDegree);
    }
}